src/RamOperation.h

```diff
@@ -63,67 +69,109 @@ class RamOperation : public RamNode {
         addCondition(std::move(c), this);
     }
 
-    /** get the optional condition on this level */
+    /** Get condition */
     RamCondition* getCondition() const {
         return condition.get();
     }
 
-    /** Obtains a list of child nodes */
+    /** Obtain list of child nodes */
     std::vector<const RamNode*> getChildNodes() const override {
         if (!condition) {
             return toVector<const RamNode*>();
         }
         return {condition.get()};
     }
+
+    /** Create clone */
+    RamOperation* clone() const override = 0;
+
+    /** Apply mapper */
+    void apply(const RamNodeMapper& map) override {
+        if (condition) {
+            condition = map(std::move(condition));
+        }
+    }
+
+protected:
+    /** Check equality */
+    bool equal(const RamNode& node) const override {
+        assert(dynamic_cast<const RamOperation*>(&node));
+        const RamOperation& other = static_cast<const RamOperation&>(node);
+        if (getCondition() != nullptr && other.getCondition() != nullptr) {
```

Comment:
I'm probably missing something, but aren't all these ifs equivalent to a single `return getCondition() == other.getCondition();`?

Is the first if body meant to be comparing pointers or objects pointed to?

Reply:
Indeed this was a semantic error. It should have been 
```
return *getCondition() == *other.getCondition();
```